lib/hostip.c

```diff
@@ -819,48 +819,95 @@ CURLcode Curl_loadhostpairs(struct Curl_easy *data)
     }
     else {
       struct Curl_dns_entry *dns;
-      Curl_addrinfo *addr;
+      Curl_addrinfo *head = 0, *tail = 0;
       char *entry_id;
       size_t entry_len;
-      char buffer[256];
-      char *address = &buffer[0];
-
-      if(3 != sscanf(hostp->data, "%255[^:]:%d:%255s", hostname, &port,
-                     address)) {
-        infof(data, "Couldn't parse CURLOPT_RESOLVE entry '%s'!\n",
-              hostp->data);
-        continue;
+      char address[64];
+      char *addrPointer;
+      char *last;
+      char *tmp;
+      unsigned long tmpPort;
+      int error = true;
+
+      tmp = strchr(hostp->data, ':');
+      if(!tmp || (tmp - hostp->data) >= (ptrdiff_t)sizeof(hostname)) {
+        goto err;
       }
+      memcpy(hostname, hostp->data, tmp - hostp->data);
+      hostname[tmp - hostp->data] = '\0';
 
-      /* allow IP(v6) address within [brackets] */
-      if(address[0] == '[') {
-        size_t alen = strlen(address);
-        if(address[alen-1] != ']')
-          /* it needs to also end with ] to be valid */
-          continue;
-        address[alen-1] = 0; /* zero terminate there */
-        address++; /* pass the open bracket */
+      tmpPort = strtoul(tmp + 1, &tmp, 10);
+      if(!tmpPort || tmpPort > USHRT_MAX || *tmp != ':' || !(tmp + 1)) {
+        goto err;
       }
+      port = (int)tmpPort;
+
+      while(true) {
+        size_t alen;
+
+        last = tmp + 1;
+        tmp = strchr(last, ',');
+        if(tmp)
+          alen = tmp - last;
+        else
+          alen = strlen(hostp->data) - (last - hostp->data);
+
+        if(alen >= sizeof(address))
+          goto err;
+
+        memcpy(address, last, alen);
+        address[alen] = '\0';
+
+        /* allow IP(v6) address within [brackets] */
+        if(address[0] == '[') {
+          if(address[alen-1] != ']')
+            /* it needs to also end with ] to be valid */
+            continue;
```

Comment:
though i'm aware this is the way it ws before should we just error here. what is the advantage to continuing if [ only, is that ever valid

Reply:
The continue is to go on to the next hostp in the the linked list of hostpairs. It didn't seem beneficial to change this behavior but I'm not attached to that.

Reply:
Ah. I see what you mean. Yes, it should be goto err. Fixing that now.

Reply:
I fixed that now. You're completely right. The continue would be for the wrong loop here.